core/src/main/java/io/grpc/internal/ServiceConfigUtil.java

```diff
@@ -0,0 +1,197 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.internal;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.base.Verify;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Random;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+/**
+ * Helper utility to work with service configs.
+ */
+final class ServiceConfigUtil {
+
+  private static final Logger logger = Logger.getLogger(ServiceConfigUtil.class.getName());
+
+  // From https://github.com/grpc/proposal/blob/master/A2-service-configs-in-dns.md
+  static final String SERVICE_CONFIG_PREFIX = "_grpc_config=";
+  private static final Set<String> SERVICE_CONFIG_CHOICE_KEYS =
+      Collections.unmodifiableSet(
+          new HashSet<String>(
+              Arrays.asList("clientLanguage", "percentage", "clientHostname", "serviceConfig")));
+
+  private ServiceConfigUtil() {}
+
+  /**
+   * Determines if a given Service Config choice applies, and if so, returns it.
+   *
+   * @see <a href="https://github.com/grpc/proposal/blob/master/A2-service-configs-in-dns.md">
+   *   Service Config in DNS</a>
+   * @param choice The service config choice.
+   * @return The service config object or {@code null} if this choice does not apply.
+   */
+  @Nullable
+  @SuppressWarnings("BetaApi") // Verify isn't all that beta
+  static Map<String, Object> maybeChooseServiceConfig(
+      Map<String, Object> choice, Random random, String hostname) {
+    for (Entry<String, ?> entry : choice.entrySet()) {
+      Verify.verify(SERVICE_CONFIG_CHOICE_KEYS.contains(entry.getKey()), "Bad key: %s", entry);
+    }
+    if (choice.containsKey("clientLanguage")) {
+      List<Object> clientLanguages = getList(choice, "clientLanguage");
+      if (clientLanguages.size() != 0) {
+        boolean javaPresent = false;
+        for (int i = 0; i < clientLanguages.size(); i++) {
+          String lang = getString(clientLanguages, i).toLowerCase(Locale.ROOT);
+          if ("java".equals(lang)) {
+            javaPresent = true;
+            break;
+          }
+        }
+        if (!javaPresent) {
+          return null;
+        }
+      }
+    }
+    if (choice.containsKey("percentage")) {
+      int pct = getDouble(choice, "percentage").intValue();
+      Verify.verify(pct >= 0 && pct <= 100, "Bad percentage", choice.get("percentage"));
+      if (pct == 0) {
+        return null;
+      } else if (pct != 100 && pct < random.nextInt(100)) {
```

Comment:
- The check for `pct != 100` is not necessary.
- 100 is excluded from possible return values of `random.nextInt(100)`, so if `pct == 99`, it will always NOT return null. So should check `pct <= random.nextInt(100)` instead.
- The check for `pct==0` is also not necessary if you simply check  `pct <= random.nextInt(100)`

```java
if (pct <= random.nextInt(100)) {
  return null;
}
```

Reply:
This calculation is incorrect, because it is off by one.   The conditions I put here are the correct ones.